rust/core-lib/src/file.rs

```diff
@@ -160,11 +213,74 @@ impl FileManager {
             return Err(FileError::HasChanged(path.to_owned()));
         } else {
             let encoding = self.file_info[&id].encoding;
+            #[cfg(feature = "notify")]
+            self.watcher.unwatch(&path, OPEN_FILE_EVENT_TOKEN);
+
             try_save(path, text, encoding).map_err(|e| FileError::Io(e, path.to_owned()))?;
             self.file_info.get_mut(&id).unwrap().mod_time = get_mod_time(path);
+
+            #[cfg(feature = "notify")]
+            self.watcher.watch(&path, false, OPEN_FILE_EVENT_TOKEN);
+
+            #[cfg(feature = "notify")]
+            self.update_current_position_in_tail(path, id)?;
         }
         Ok(())
     }
+
+    #[cfg(feature = "notify")]
+    pub fn update_current_position_in_tail(
+        &mut self,
+        path: &Path,
+        id: BufferId,
+    ) -> Result<(), FileError> {
+        let existing_file_info = self.file_info.get_mut(&id).unwrap();
+
+        let mut f = File::open(path).map_err(|e| FileError::Io(e, path.to_owned()))?;
+        let end_position =
+            f.seek(SeekFrom::End(0)).map_err(|e| FileError::Io(e, path.to_owned()))?;
+
+        existing_file_info.tail_details.current_position_in_tail = end_position;
+        Ok(())
+    }
+}
+
+/// When tailing a file, instead of reading file from beginning, we need to get changes from the end.
+/// This method does that.
+#[cfg(feature = "notify")]
+fn try_tailing_file<P>(path: P, current_position: u64) -> Result<(Rope, FileInfo), FileError>
+where
+    P: AsRef<Path>,
+{
+    let mut f =
+        File::open(path.as_ref()).map_err(|e| FileError::Io(e, path.as_ref().to_owned()))?;
+    let end_position =
+        f.seek(SeekFrom::End(0)).map_err(|e| FileError::Io(e, path.as_ref().to_owned()))?;
+
+    let diff = end_position - current_position;
+    let mut buf = vec![0; diff as usize];
+    f.seek(SeekFrom::Current(-(buf.len() as i64))).unwrap();
```

Comment:
this could fail; we should handle the error.

Reply:
Agree.